src/index.js

```diff
@@ -0,0 +1,235 @@
+ *
+ * Signature of the progress callback is: function(migrationObject: object, currentMigrationNumber: int, totalMigrationsCount: int)
+ *
+ * @param {string} path - Path to initialized (!) JS-IPFS repo
+ * @param {Object} options - Options for migration
+ * @param {int?} options.toVersion - Version to which the repo should be migrated, if undefined repo will be migrated to the latest version.
+ * @param {boolean?} options.ignoreLock - Won't lock the repo for applying the migrations. Use with caution.
+ * @param {object?} options.repoOptions - Options that are passed to migrations, that can use them to correctly construct datastore. Options are same like for IPFSRepo.
+ * @param {function?} options.onProgress - Callback which will be called after each executed migration to report progress
+ * @param {boolean?} options.isDryRun - Allows to simulate the execution of the migrations without any effect.
+ * @param {array?} options.migrations - Array of migrations to migrate. If undefined, the bundled migrations are used. Mainly for testing purpose.
+ * @returns {Promise<void>}
+ */
+async function migrate (path, { toVersion, ignoreLock = false, repoOptions, onProgress, isDryRun = false, migrations }) {
+  migrations = migrations || defaultMigrations
+
+  if (!path) {
+    throw new errors.RequiredParameterError('Path argument is required!')
+  }
+
+  if (!(await repoInit.isRepoInitialized(path))) {
+    throw new errors.NotInitializedRepoError(`Repo in path ${path} is not initialized!`)
+  }
+
+  if (toVersion && (!Number.isInteger(toVersion) || toVersion <= 0)) {
+    throw new errors.InvalidValueError('Version has to be positive integer!')
+  }
+  toVersion = toVersion || getLatestMigrationVersion(migrations)
+
+  if (toVersion > getLatestMigrationVersion(migrations)) {
+    throw new errors.InvalidValueError('The ipfs-repo-migrations package does not have migration for version: ' + toVersion)
+  }
+
+  const currentVersion = await repoVersion.getVersion(path)
+
+  if (currentVersion === toVersion) {
+    log('Nothing to migrate, skipping migrations.')
+    return
+  }
+
+  if (currentVersion > toVersion) {
+    log(`Current repo's version (${currentVersion}) is higher then toVersion (${toVersion}), nothing to migrate.`)
+    return
+  }
+
+  let lock
+  if (!isDryRun && !ignoreLock) lock = await repoLock.lock(currentVersion, path)
+
+  try {
+    let counter = 0
+    let totalMigrations = toVersion - currentVersion
+    for (let migration of migrations) {
+      if (toVersion !== undefined && migration.version > toVersion) {
+        break
+      }
+      if (migration.version <= currentVersion) {
+        continue
+      }
+
+      counter++
+      log(`Migrating version ${migration.version}`)
+      try {
+        if (!isDryRun) await migration.migrate(path, repoOptions, isBrowser)
+      } catch (e) {
+        e.message = `During migration to version ${migration.version} exception was raised: ${e.message}`
```

Comment:
Can we use [explain-error](https://www.npmjs.com/package/explain-error) here so as not to lose the stack trace?

Reply:
Well, of course, we can, but I am not sure how much of added value will it bring. The stack traces are persisted using this approach.
I looked into how `explain-error` generates stack traces and at least to me, it seems super confusing. Bellow is how stack trace looks like with this approach:
```
  repo-migrations:repo_fs_lock locking /Users/adam/.jsipfs/repo.lock +0ms
  repo-migrations:migrator Migrating version 8 +0ms
  repo-migrations:repo_fs_lock releasing lock /Users/adam/.jsipfs/repo.lock +4ms
  repo-migrations:cli {
  repo-migrations:cli   error: Error: During migration to version 8 exception was raised: Some err
  repo-migrations:cli       at Object.migrate (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/migrations/migration-8/index.js:55:9)
  repo-migrations:cli       at Object.migrate (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/src/index.js:99:40)
  repo-migrations:cli       at async migrate (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/src/commands.js:67:5)
  repo-migrations:cli       at async main (/Users/adam/Projects/ipfs/js-ipfs-repo-migrations/src/cli.js:52:22),
  repo-migrations:cli   argv: {
  repo-migrations:cli     _: [ 'migrate' ],
  repo-migrations:cli     resolve: [Function],
  repo-migrations:cli     reject: [Function],
  repo-migrations:cli     '$0': 'src/cli.js'
  repo-migrations:cli   }
  repo-migrations:cli } +0ms
During migration to version 8 exception was raised: Some err
```

Reply:
Ok if you think the error will be clear enough let's leave it 👍